fix: JOIN-31701 update pubsub library to publish and consume avro encoded messages

eugeneswave opened the pull request:
Code doesn't look really beautiful right now because of the try catches, to update the schema, if message parse fails, but we can remove it later, when we move to the AVRO completely.
I left TODO for it. Let's also merge it after the check on a small service (CV-Parser) and after adding https://joinsolutionsag.atlassian.net/browse/JOIN-31233

ddadonJoin commented on packages/pubsub/src/Subscriber.ts:
```diff
@@ -122,15 +128,42 @@ export class Subscriber<T = unknown> {
+        //reload the topic to correctly process next message, if there was no schema
+        if (!this.avroType) {
+          this.topic = this.client.topic(this.topicName)
+          // we have to do it this way because processMsg is not async in our library
+          // and if we add async everywhere we will have to change current library API
```
Maybe explain that this will only happen when we migrate from JSON to Avro and that no message will be lost

ddadonJoin commented on packages/pubsub/src/TopicHandler.ts:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  protected async getTopicType() {
+    // const schemaName = topic.metadata?.schemaSettings?.schema
```
You can remove this

ddadonJoin commented on packages/pubsub/src/TopicHandler.ts:
```diff
@@ -0,0 +1,30 @@
+    this.topic = client.topic(topicName)
+  }
+
+  protected async getTopicType() {
```
Can you put an explicit return type? Also This should have been raised by the linter so could you also update the linter?

ddadonJoin commented on packages/pubsub/src/TopicHandler.ts:
```diff
@@ -0,0 +1,30 @@
+  protected async getTopicType() {
+    // const schemaName = topic.metadata?.schemaSettings?.schema
+    const metadata = await this.topic.getMetadata()
+    const schemaName = metadata?.[0].schemaSettings?.schema
```
What happens if `[0]` return undefined? This could raise an exception no?

ddadonJoin commented on packages/pubsub/src/TopicHandler.ts:
```diff
@@ -0,0 +1,30 @@
+import { PubSub, Topic } from '@google-cloud/pubsub'
+import { Schema } from 'avsc'
+import * as avro from 'avsc'
```
We should avoid star import I think :)

ddadonJoin reviewed (changes requested):
Good start :)

ddadonJoin reviewed (changes requested):
Good start :)

ddadonJoin reviewed (changes requested):
Good start :)

kirpichenko commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -1,26 +1,47 @@
+        //it's a corner case when application started without topic schema, and then schema was added to the topic
+        //in this case we are trying to get again topic data and resend with the schema if it's appeared
+        this.topic = this.client.topic(this.topic.name)
+        this.avroType = await this.getTopicType()
```
shouldn't the schema change happen together with deployment?

kirpichenko replied:
maybe we can continue publishing as a json if the schema is not set, additionally pass the schema/type via attributes to the client and let the client decide how to parse the message?

eugeneswave replied:
> shouldn't the schema change happen together with deployment?

Yes, as we are already managing topics in the code, so we will assign schema to the topic in the code, and we will have schema update as a part of deployment (probably gcloud calls inside docker image or terraform if we can manage it). 

The problem is, we can not completely stop the server, and in this case some pods will be still not updated, and some pods will be new.

So new pods will work fine, but old pods will publish the data in JSON, so they will receive error, then they will get new schema from the topic, and send the data with the schema, and that's why we have this try-catch here

eugeneswave replied:
> maybe we can continue publishing as a json if the schema is not set, additionally pass the schema/type via attributes to the client and let the client decide how to parse the message?

Our end goal is to have avro enabled on the topic, to use pubsub subscriber for BigQuery, and I don't see how additional field in json will help us here. Because for now we will need to handle 2 cases JSON or AVRO, with additional field, it will be 3 cases, and I don't see how it helps us.

kirpichenko commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -37,4 +58,17 @@ export class Publisher<T = unknown> {
+    // for now we are checking that it's not null before calling sendAvroMessage
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const buffer = this.avroType!.toBuffer(data)
+    await this.publishWithLog({ data: buffer })
```
the logger seems like will print a buffer instead of the published message

ddadonJoin commented on packages/pubsub/src/DataParser.ts:
```diff
@@ -9,4 +9,21 @@ export class DataParser {
+  private deepReplaceInObject(obj: Record<string, unknown>)  {
+    for (const key in obj) {
+      if (obj[key] === null) {
+        obj[key] = undefined;
```
Our code style does not use semicolons, the linter should have catch it though, can you check why?

eugeneswave replied:
Looks like we don't have this rule enable in the linter, but of course removing it here

ddadonJoin commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -1,26 +1,46 @@
+      } catch (e) {
+        //it's a corner case when application started without topic schema, and then schema was added to the topic
+        //in this case we are trying to get again topic data and resend with the schema if it's appeared
+        this.topic = this.client.topic(this.topic.name)
```
Why do we need to re-assign topic here?

eugeneswave replied:
My logic was, because getTopicType works for the topic already received, and we need to update the topic first, and the execute getTopicType. Bus as we are having callback for getting topic metadata, maybe we will always have a fresh one, and we don't need to refresh topic. Checking this.

eugeneswave replied:
yeah, it works, thanks for the comment! This allows us to skip refreshing the topic

ddadonJoin commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -1,26 +1,46 @@
   public async publishMsg(data: T): Promise<void> {
-    const messageId = await this.topic.publishMessage({ json: data })
-    this.logger?.info(`PubSub: Message sent for topic: ${this.topicName}:`, { data, messageId })
+    // Later we want to have only topic with specified schema and remove if block below
```
Can you add a TODO prefix in the comment to make the intent of removing it later clearer?

ddadonJoin commented on packages/pubsub/src/TopicHandler.ts:
```diff
@@ -0,0 +1,29 @@
+    this.topic = client.topic(topicName)
+  }
+
+  protected async getTopicType(): Promise<Type | undefined> {
```
I know this code predate our discussion from yesterday but we want to use the writer's schema over there, we can sync to align on how you can get it if you want :)

ddadonJoin commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -37,4 +57,19 @@ export class Publisher<T = unknown> {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const buffer = this.avroType!.toBuffer(data)
+    const messageId = await this.topic.publishMessage({ data: buffer })
+    this.logger?.info(`PubSub: Message sent for topic: ${this.topicName}:`, { data, messageId })
+
+  }
+
+  private async sendJsonMessage(message: MessageOptions) {
+    const messageId = await this.topic.publishMessage(message)
+    this.logger?.info(`PubSub: Message sent for topic: ${this.topicName}:`, { message, messageId })
```
Can you add the encoding scheme on the log to make it easier to search? Like `Avro-Message sent for topic...` and `Json-Message sent for topic...`

I could even give you the avro schema fingerprint if you want to log it too

ddadonJoin replied:
Yes good idea to get the schema version id in the logs

ddadonJoin commented on packages/pubsub/src/Subscriber.ts:
```diff
@@ -122,15 +138,34 @@ export class Subscriber<T = unknown> {
-    const dataParsed = dataParser.parse(message.data) as T
+    if (this.avroType) {
+      const dataParsedWithNulls = this.avroType.fromBuffer(message.data) as T
+      dataParsed = dataParser.replaceNullsWithUndefined(dataParsedWithNulls)
```
Nice :), would it be possible to include this as a custom deserializer like `DateType` and register it in `Type.forSchema()` below? It would encapsulate the logic even more

eugeneswave replied:
Couldn't find any way to do this for the whole conversion. Yes it works for logical types, but not for the whole object, and every type.
https://github.com/mtth/avsc/wiki/API#typeforschemaschema-opts

ddadonJoin commented on packages/pubsub/src/Subscriber.ts:
```diff
@@ -122,15 +138,34 @@ export class Subscriber<T = unknown> {
+        if (this.avroType) {
+         throw e
+        }
+        this.topic = this.client.topic(this.topicName)
+        this.avroType = await this.getTopicType()
+        dataParsed = this.parseData(message)
```
There is no `return` on line 158 so does that mean we will call this block every time we process a message? Also that mean we will call this.parseData() twice

kirpichenko commented on packages/pubsub/src/DataParser.ts:
```diff
@@ -9,4 +9,21 @@ export class DataParser {
+        obj[key] = undefined
+      } else if (Array.isArray(obj[key])) {
+        (obj[key] as Record<string, unknown>[]).forEach(member => this.deepReplaceInObject(member))
+      } else if (typeof obj[key] === 'object') {
```
you need to be careful with this check, bc Date is also object

eugeneswave replied:
added additional check if object has a property

kirpichenko commented on packages/pubsub/src/Publisher.ts:
```diff
@@ -1,26 +1,47 @@
   public async initialize() {
     try {
       await this.initializeTopic()
+      this.avroType = await this.getTopicType()
```
I did not get it. I mean that instead of inheritance it's enough to use composition. You can keep the same class. I just don't see much sense here and prefer to avoid inheritance when possible

eugeneswave replied:
yeah, sorry for misunderstanding, I updated code base, it should look better now

kirpichenko commented on packages/pubsub/src/Subscriber.ts:
```diff
@@ -80,22 +79,39 @@ export class Subscriber<T = unknown> {
     }
   }
 
+  private readonly deadLetterSubscriptionName?: string
```
why did you move it? we usually keep all class properties on the top

kirpichenko commented on packages/pubsub/src/Subscriber.ts:
```diff
@@ -122,15 +139,33 @@ export class Subscriber<T = unknown> {
   private processMsg(asyncCallback: (msg: IParsedMessage<T>) => Promise<void>) {
-    return (message: Message) => {
-      const dataParsed = this.parseData(message)
+    return async (message: Message) => {
```
i think you can do smth like instead of callbackify, bc it looks extremely confusing 
```
return  (message: Message) => new Promise(async(resolve, reject) => {
   ... await ...
})
```

eugeneswave commented on packages/pubsub/src/DataParser.ts:
```diff
@@ -9,4 +9,21 @@ export class DataParser {
+      if (obj[key] === null) {
+        obj[key] = undefined
+      } else if (Array.isArray(obj[key])) {
+        (obj[key] as Record<string, unknown>[]).forEach(member => this.deepReplaceInObject(member))
```
added additional check if object has a property

eugeneswave commented:
I've addressed the issues except callbackify, will check it tomorrow, need more time, and also added 
https://joinsolutionsag.atlassian.net/browse/JOIN-31233
